src/mod_webdav.c

```diff
@@ -1796,6 +1808,16 @@ static handler_t mod_webdav_put(server *srv, connection *con, plugin_data *p, ha
 				return HANDLER_FINISHED;
 			}
 
+			if (length >= 0) {
```

Comment:
Do not trust length from header blindly.

The uploaded content length (chunkqueue_length(cq)) (and cq is con->request_content_queue in this routine) should match the size of (end - offset) if end is set (scoped above), or (length - offset), and it should be an error if the length of content does not match.  (unless length = -1)

If end is provided and is different from length, then length should be checked against existing file size and it should be an error if the length does not match the current file length.

Since we are about to write the contents to the file, and I have added a requirement that length match the file size or about to be written to that file size, what is the benefit of the ftruncate()? 

(I am out of time at the moment.   I'll have to review the rest some other time.)

Reply:
It does actually try to provide a way to truncate files remotely using the Content-Range scheme, but nah